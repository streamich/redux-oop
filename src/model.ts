import {Connector, getConnector} from './atoms';
import {TId} from './types';

export type TObject = {[key: string]: any};

export class Model {
    static cid = 1;

    /**
     * A special property of models, the cid or client id is a unique
     * identifier automatically assigned to all models when they're first created
     * @type {number}
     */
    cid = Model.cid++;

    /**
     * Connector set on the model using `@id` decorator.
     * @returns {Connector}
     * @private
     */
    get __$$$__(): Connector {
        return getConnector(this);
    }

    /**
     * Get a single key from `attributes` object.
     * @param {string} key
     * @returns {any}
     */
    get(key: string): any {
        return this.__$$$__.select()[key];
    }

    /**
     * Update a single key in `attributes` object.
     * @param {string} key
     * @param value
     */
    set(key: string, value: any) {
        this.patch({[key]: value});
    }

    /**
     * Update model's `attributes` by merging in supplied object.
     * @param {TObject} patch
     */
    patch(patch: TObject) {
        this.__$$$__.patch(patch);
    }

    /**
     * The attributes property is the internal hash containing the
     * model's state in Redux.
     *
     * Please don't modify the `.attributes` object directly. If you'd
     * like to retrieve and munge a copy of the model's attributes,
     * use `.toJSON()` instead.
     * @returns {TObject}
     */
    get attributes(): TObject {
        return this.__$$$__.select();
    }

    /**
     * Return a shallow copy of the model's `attributes`.
     * @returns {TObject}
     */
    toJSON(): TObject {
        return {...this.attributes};
    }

    /**
     * Returns `true` if the attribute is set to a non-null or non-undefined value.
     * @param {string} key
     * @returns {boolean}
     */
    has(key: string): boolean {
        const value = this[key];

        return value !== null && value !== void 0;
    }

    /**
     * Remove an attribute by deleting it from the internal attributes hash.
     * @param {string} key
     */
    unset(key: string) {
        this.__$$$__.del([...this.path(), key]);
    }

    /**
     * Returns a path array to the current model.
     * @returns {string[]}
     */
    path(): string[] {
        return this.__$$$__.path();
    }

    /**
     * Removes all attributes from the model, including the `id` attribute.
     */
    clear() {
        this.__$$$__.del(this.path());
    }

    /**
     * Returns a new instance of the model with identical attributes.
     * @param {TId} id
     * @returns {this}
     */
    clone(id: TId): this {
        const {idKey} = this.__$$$__;
        if (!idKey) throw Error('Cannot clone model without @id.');

        const Klass = this.constructor as any;
        const instance = new Klass();
        const {[idKey]: omit, ...attributes} = this.attributes;
        instance[idKey] = id;
        instance.patch(attributes);
        return instance;
    }
}
