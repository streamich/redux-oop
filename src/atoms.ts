import {Store} from 'redux';
import {pathDelete, pathPatch} from './actions';
import {
    sConnector,
    sId,
    sLocator,
    sParent,
    sPath,
    sStore,
    sTable,
    sym,
} from './util';
import {TId} from './types';
import {genPathSelector, selectByPath} from './selectors';

export type TModelLocator = (table: string, id: TId) => string[];
const defaultModelLocator: TModelLocator = (table, id) => [table, String(id)];

const setStore = (obj, store: () => Store<any>) => {
    obj[sStore] = store;
};

const setPath = (obj, path: () => string[]) => {
    obj[sPath] = path;
};

const getStore = obj => {
    return obj[sStore]();
};

const getPath = obj => {
    return obj[sPath]();
};

const hidden = (obj, key, value) =>
    Object.defineProperty(obj, key, {
        enumerable: false,
        writable: true,
        value,
    });

export const setConnector = (obj): Connector => {
    const conn = new Connector();
    conn.obj = obj;
    hidden(obj, sConnector, conn);
    obj[sConnector] = conn;
    return conn;
};

export const getConnector = (obj): Connector => {
    if (!obj || typeof obj !== 'object') return null;
    if (!obj[sConnector]) return setConnector(obj);
    else return obj[sConnector];
};

export class Connector {
    idKey: string;
    obj: any;
    parent: Connector = null;
    store: Store<any> = null;
    table: string;
    locator: TModelLocator;
    key: string;
    selector: (state) => any;

    get Store(): Store<any> {
        return this.store || (this.parent && this.parent.Store);
    }

    set Store(store: Store<any>) {
        this.store = store;
    }

    get Id(): TId {
        return this.obj[sId];
    }

    set Id(id: TId) {
        hidden(this.obj, sId, id);
        this.patch({[this.idKey]: id});
    }

    path(): string[] {
        if (this.locator) {
            return this.locator(this.table, this.obj[sId]);
        } else if (this.parent) {
            return [...this.parent.path(), this.key];
        } else {
            throw Error('No path set');
        }
    }

    get Selector(): (state) => any {
        if (!this.selector) this.selector = genPathSelector(this.path());
        return this.selector;
    }

    select() {
        return this.Selector(this.Store.getState());
    }

    dispatch(action) {
        this.Store.dispatch(action);
    }

    patch(patch: {[key: string]: any}) {
        this.dispatch(pathPatch(this.path(), patch));
    }

    del(path: string[]) {
        this.dispatch(pathDelete(path));
    }
}

export function id(
    store: Store<any>,
    table: string,
    locator: TModelLocator = defaultModelLocator
) {
    return (obj, key: string) => {
        const conn = setConnector(obj);
        conn.Store = store;
        conn.table = table;
        conn.locator = locator;
        conn.idKey = key;
        Object.defineProperties(obj, {
            [key]: {
                get: (): TId => conn.Id,
                set: (id: TId) => (conn.Id = id),
            },
        });
    };
}

export function prop(obj, key: string) {
    const conn = getConnector(obj);
    Object.defineProperty(obj, key, {
        enumerable: true,
        set: value => {
            let propConnector = getConnector(value);
            const sProp = sym('.' + key);

            if (propConnector) {
                hidden(obj, sProp, value);
                propConnector.key = key;
                propConnector.parent = conn;
            } else {
                const propObj = obj[sProp];
                if (typeof propObj !== void 0) delete obj[sProp];
                if (value !== void 0) conn.patch({[key]: value});
                else {
                    conn.Store.dispatch(pathDelete([...conn.path(), key]));
                }
            }
        },
        get: () => {
            const sProp = sym('.' + key);
            const value = obj[sProp];
            if (value !== void 0) {
                return value;
            } else {
                return conn.select()[key];
            }
        },
    });
}
