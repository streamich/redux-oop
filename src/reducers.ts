import {MODEL_PATCH, PATH_DELETE, PATH_PATCH} from './actions';

export function reducePathPatch(state, {path, patch}) {
    if (!path.length) return {...state, ...patch};
    else {
        const [step, ...restPath] = path;
        return {
            ...state,
            [step]: reducePathPatch(state[step] || {}, {path: restPath, patch}),
        };
    }
}

export function reducePathDelete(state, {path}) {
    if (path.length === 1) {
        const [step] = path;
        const {[step]: omit, ...newState} = state;
        return newState;
    } else {
        const [step, ...restPath] = path;
        return {
            ...state,
            [step]: reducePathDelete(state[step] || {}, {path: restPath}),
        };
    }
}

export function reducePatch(state, {table, id, patch}) {
    return reducePathPatch(state, {path: [table, id], patch});
}

export function reduce(state, action) {
    switch (action.type) {
        case PATH_PATCH:
            return reducePathPatch(state, action);
        case PATH_DELETE:
            return reducePathDelete(state, action);
        case MODEL_PATCH:
            return reducePatch(state, action);
        default:
            return state;
    }
}
