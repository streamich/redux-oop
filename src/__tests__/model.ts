import {Model} from '../model';
import {id, prop} from '../atoms';
import {createStore} from 'redux';
import {reduce} from '../reducers';

describe('model', () => {
    describe('@prop', () => {
        it('is a decorator', () => {
            expect(typeof prop).toBe('function');
            expect(prop.length).toBe(2);
        });
    });
    describe('new Model', () => {
        it('creates model without crashing', () => {
            const model = new Model();
        });
    });
    describe('.set', () => {
        it('set a single key', () => {
            const store = createStore(reduce, {});
            class MyModel extends Model {
                @id(store, 'models')
                id = 1;
            }
            const model = new MyModel();
            model.set('foo', 'bar');
            expect(store.getState()).toMatchSnapshot();
        });
    });
});
