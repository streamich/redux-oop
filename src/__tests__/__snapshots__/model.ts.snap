// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model .set set a single key 1`] = `
Object {
  "models": Object {
    "1": Object {
      "foo": "bar",
      "id": 1,
    },
  },
}
`;
